drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+        let mut state = self.state.write();
+        let mut overrun_seen = false;
+
+        // Set to true when `tx_iter` has been exhausted. Note that we may
+        // continue sending `0` after this if we are still clocking in bytes
+        // from the SP.
+        let mut response_sent = false;
 
-                // Track Spi Select Asserted and Deasserted to determine if
-                // we are in frame and update Rx/Tx state as needed.
-                if intstat.ssa().bit() {
-                    self.spi.ssa_clear();
-                    inframe = true;
+        loop {
+            let fifostat = self.spi.fifostat();
+
+            // We check for CSn deasserted at the beginning of the loop because
+            // we always want an opportunity drain our fifos after we see the bit set.
+            let csn_deasserted = self.spi.ssd();
+
+            if fifostat.txerr().bit() {
+                // We missed our window to send :(
+                self.spi.txerr_clear();
+                self.stats.tx_underrun = self.stats.tx_underrun.wrapping_add(1);
+                // If we've already written our full response, underrun doesn't matter.
+                // This can happen if we clock out BUF_SIZE of data, but are waiting
+                // for a CSn de-assert.
+                if state != WriteState::ResponseWritten {
+                    state = WriteState::Underrun;
+                }
+            }
+
+            if fifostat.rxerr().bit() {
+                // We missed a byte to read. But we are only expecting 0s, so
+                // this isn't critical, except to judge if we got a CSn pulse
+                // or not.
+                self.spi.rxerr_clear();
+                self.stats.rx_overrun = self.stats.rx_overrun.wrapping_add(1);
+                overrun_seen = true;
+            }
+
+            let mut io = true;
+            while io == true {
+                io = false;
+                if self.spi.can_tx() {
+                    io = true;
+                    if let Some(b) = tx_iter.next() {
+                        self.spi.send_u8(b);
+                    } else {
+                        response_sent = true;
+                        // Just clock out zeros and prevent an unnecessary underrun
+                        self.spi.send_u8(0);
+                    }
                 }
 
-                // Note that while Tx is done at end of frame,
-                // FIFORD may still have bytes to read.
-                if intstat.ssd().bit() {
-                    self.spi.ssd_clear();
-                    if transmit {
-                        self.gpio.set_val(ROT_IRQ, Value::One).unwrap_lite();
+                if self.spi.has_byte() {
+                    io = true;
+                    // The SP should only clock out zeros while waiting for a
+                    // response. We'll record it, but still finish sending our reply.
+                    if self.spi.read_u8() != 0 {
```

Comment:
We've talked about this already.  It is valid for the SP to forget everything ant start a new message or to send continuous messages and take the responses to previous messages while sending the next.
This code is mixing handler state into the data link.

Reply:
The SP can certainly do whatever it wants, and that doesn't really change here, other than logging. 

I think there are practical problems with this strategy of letting the SP do what it wants with no coordination though. The RoT code as structured here, and on master, is only written to understand and be able to process one request from the SP at a time. Buffers are sized around handling a maximum sized single message, and every time through the `write` state machine here, or `pio` loop on master, we clear reset the writer to start from the beginning of the tx buffer. If we wanted to allow pipelining of requests we'd have to append them to the rx buffer, and we'd have to use the tx buffer in some ring fashion. Even then, however, we'd need to set some maximum number of messages from the SP that we could allow to be pipelined so we could appropriately size the buffer. This code explicitly encodes the defacto choice on master and limits the SP to sending one request at a time, which makes the code path for the RoT simpler.

The SP can still choose to abandon the prior request at any time, ignore responses, etc..., but the RoT still has to understand what to do in this case. The RoT knows the limits of its buffers, and also knows that it signals a reply asserting `ROT_IRQ`. If the RoT is responding, it does have room to handle one more concurrent message of maximum size in its RX buffer, but not more than that. If the RoT already has part of a message in its rx buffer, it cannot handle a full message, and so it will be forced to either drop bytes, or clear its own RX buffer at the beginning of this call and prepare to receive the next message. This seems more complex than the strategy of just expecting the SP to send one message at a time, and not worry too much about multiple messages in a buffer.

Reply:
Changed, with the buffer changes mentioned in another comment.